packages/cli/src/ceramic-daemon.ts

```diff
@@ -307,154 +189,105 @@ class CeramicDaemon {
-      res.json({ docId: doctype.id.toString(), state: DoctypeUtils.serializeState(doctype.state) })
-    } catch (e) {
-      return next(e)
-    }
-    next()
+    const doctype = await this.ceramic.applyCommit(docId, DoctypeUtils.deserializeCommit(commit), docOpts)
+    res.json({ docId: doctype.id.toString(), state: DoctypeUtils.serializeState(doctype.state) })
   }
 
   /**
    * Load multiple documents and paths using an array of multiqueries
    */
-  async multiQuery (req: Request, res: Response, next: NextFunction): Promise<void> {
+  async multiQuery (req: Request, res: Response): Promise<void> {
     const { queries } = <MultiQueries> req.body
-    try {
-      const results = await this.ceramic.multiQuery(queries)
-      const response = Object.entries(results).reduce((acc, e) => {
-        const [k, v] = e
-        acc[k] = DoctypeUtils.serializeState(v.state)
-        return acc
-      }, {})
-      res.json(response)
-    } catch (e) {
-      return next(e)
-    }
-    next()
+    const results = await this.ceramic.multiQuery(queries)
+    const response = Object.entries(results).reduce((acc, e) => {
+      const [k, v] = e
+      acc[k] = DoctypeUtils.serializeState(v.state)
+      return acc
+    }, {})
+    res.json(response)
   }
 
   /**
    * Pin document
    */
-  async pinDocument (req: Request, res: Response, next: NextFunction): Promise<void> {
+  async pinDocument (req: Request, res: Response): Promise<void> {
     const docId = DocID.fromString(req.params.docid)
-    try {
-      await this.ceramic.pin.add(docId)
-      res.json({ docId: docId.toString(), isPinned: true })
-    } catch (e) {
-      return next(e)
-    }
-    next()
+    await this.ceramic.pin.add(docId)
+    res.json({ docId: docId.toString(), isPinned: true })
   }
 
   /**
    * Unpin document
    */
-  async unpinDocument (req: Request, res: Response, next: NextFunction): Promise<void> {
+  async unpinDocument (req: Request, res: Response): Promise<void> {
     const docId = DocID.fromString(req.params.docid)
-    try {
-      await this.ceramic.pin.rm(docId)
-      res.json({ docId: docId.toString(), isPinned: false })
-    } catch (e) {
-      return next(e)
-    }
-    next()
+    await this.ceramic.pin.rm(docId)
+    res.json({ docId: docId.toString(), isPinned: false })
   }
 
   /**
    * List pinned documents
    */
-  async listPinned (req: Request, res: Response, next: NextFunction): Promise<void> {
+  async listPinned (req: Request, res: Response): Promise<void> {
     let docId: DocID;
     if (req.params.docid) {
       docId = DocID.fromString(req.params.docid)
     }
-    try {
-      const pinnedDocIds = []
-      const iterator = await this.ceramic.pin.ls(docId)
-      for await (const id of iterator) {
-        pinnedDocIds.push(id)
-      }
-      res.json({ pinnedDocIds: pinnedDocIds })
-    } catch (e) {
-      return next(e)
+    const pinnedDocIds = []
+    const iterator = await this.ceramic.pin.ls(docId)
```

Comment:
Can we send 400 if we don't have the docId? Maybe just add an else statement above

Reply:
Actually, this is a perfect example to take a strong stance "no sanitation at http daemon level". `pin.ls` accepts both docId and `undefined`. Both are valid. The latter means "list everything pinned".

Here, the only problematic case is invalid DocID passed. This is handled in `DocID.fromString` which errors if the input can not be deserialised into `DocID`.

`loadDocumentCommits` AKA [`commits`](https://github.com/ceramicnetwork/js-ceramic/pull/1210/files/436c246640c917e21ada94412cfa3ff96840af68#diff-29a806cda83d5289fde2f9b08500d1f2335980307791376095ed3d7c600da98eR210) is a bit more interesting. It tries to parse DocID in the Express handler, but the underlying code also tries to perform deserialisation. Effectively, either DocID deserialization in [`commits` handler](https://github.com/ceramicnetwork/js-ceramic/pull/1210/files/436c246640c917e21ada94412cfa3ff96840af68#diff-29a806cda83d5289fde2f9b08500d1f2335980307791376095ed3d7c600da98eR211) or DocID deserialization inside `loadDocumentCommits` is redundant.

What makes sense to me is a proper serialization/deserialization mechanism to strictly delineate formatting (400) errors from other ones while reducing redundancy with the underlying code. The _other errors_ could be mapped to whatever http errors, including 400. As discussed, first step on the way there is getting that mapping done. Then, few discussions have to be made with regards to serialization/deserialization tooling. It better be a single thing, like a codec, it better be composable, otherwise it introduces maintenance headache. But, it is a different story from the one we have now.

Reply:
Ok so docid here is optional. Wasn't clear to me. I would prefer being more verbose. Such as `pin.ls(docId && docId || undefined)`. Or at least something so it is clear that docId is optional here.

Agree on the mapping and the fact that it is out of scope here.